src/main/kotlin/com/getkeepsafe/dexcount/DexFile.kt

```diff
@@ -0,0 +1,263 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.getkeepsafe.dexcount
+
+import com.android.dexdeps.DexData
+import com.android.dexdeps.DexDataException
+import com.android.dexdeps.FieldRef
+import com.android.dexdeps.MethodRef
+
+import java.io.File
+import java.io.RandomAccessFile
+
+import java.util.zip.ZipEntry
+import java.util.zip.ZipException
+import java.util.zip.ZipFile
+
+import java.io.EOFException
+
+/**
+ * A physical file and the {@link DexData} contained therein.
+ *
+ * A DexFile contains an open file, possibly a temp file.  When consumers are
+ * finished with the DexFile, it should be cleaned up with
+ * {@link DexFile#dispose()}.
+ */
+internal class DexFile private constructor(
+        private val file: File,
+        private val isTemp: Boolean,
+        val isInstantRun: Boolean = false) {
+
+    private val raf: RandomAccessFile = RandomAccessFile(file, "r")
+    private val data: DexData = DexData(raf)
+
+    val methodRefs: List<MethodRef>
+        get() = data.methodRefs.toList()
+
+    val fieldRefs: List<FieldRef>
+        get() = data.fieldRefs.toList()
+
+    init {
+        try {
+            data.load()
+        } catch (e: EOFException) {
+            throw DexCountException("Error loading dex file", e)
+        } catch (e: DexDataException) {
+            throw DexCountException("Error loading dex file", e)
+        }
+    }
+
+    fun dispose() {
+        raf.close()
+        if (isTemp) {
+            file.delete()
+        }
+    }
+
+    companion object {
+        /**
+         * Extracts a list of {@link DexFile} instances from the given file.
+         *
+         * DexFiles can be extracted either from an Android APK file, or from a raw
+         * {@code classes.dex} file.
+         *
+         * @param file the APK or dex file.
+         * @param dxTimeoutSecs timeout when running Dx in seconds
+         * @return a list of DexFile objects representing data in the given file.
+         */
+        @JvmStatic
+        fun extractDexData(file: File?, dxTimeoutSecs: Int): List<DexFile>
+        {
+            if (file == null || !file.exists()) {
+                return emptyList()
+            }
+
+            // AAR files need special treatment
+            if (file.name.endsWith(".aar")) {
+                return emptyList()
+                //return extractDexFromAar(file, dxTimeoutSecs)
+            }
+
+            try {
+                return extractDexFromZip(file)
+            } catch (ignored: ZipException) {
+                // not a zip, no problem
+            }
+
+            return listOf(DexFile(file, false))
+        }
+
+//        fun extractDexFromAar(file: File, dxTimeoutSecs: Int): List<DexFile>
```

Comment:
Correct - this is the part where I haven't figured out how to replicate some Groovy built-ins.